File: level-ii/question-i/cellular_automata.py
from time import time
import numpy as np
from numba import njit
from multiprocessing import Pool


class CellularAutomata:
    def __init__(self, filepath):
        self.filepath = filepath
        self.matrix = np.array([])
        self.row_count = 0
        self.column_count = 0
        self.generation = 0
        self.load_matrix()

    def attribute_next_generation(self):
        self.matrix = self.compute_next_generation()
        self.generation += 1

    def compute_next_generation(self):
        with Pool(processes=4) as pool:
            chunked_matrices = np.array_split(self.matrix, 4)
            next_generation_matrices = pool.map(self.compute_next_generation_chunk, chunked_matrices)
            return np.concatenate(next_generation_matrices)

    @staticmethod
    @njit(fastmath=True)
    def compute_next_generation_chunk(matrix):
        next_generation_matrix = np.zeros(shape=matrix.shape, dtype=np.int32)

        for i in range(matrix.shape[0]):
            for j in range(matrix.shape[1]):
                cell = matrix[i, j]
                neighbors = CellularAutomata.get_cell_neighbors(matrix, i, j)

                if cell == 0 and (neighbors > 1 and neighbors < 5):
                    next_generation_matrix[i, j] = 1
                elif cell == 1:
                    if neighbors > 3 and neighbors < 6:
                        next_generation_matrix[i, j] = 1

        return next_generation_matrix

    @staticmethod
    @njit(fastmath=True)
    def get_cell_neighbors(matrix, i, j):
        neighbors = 0

        for x in range(-1, 2):
            for y in range(-1, 2):
                if x == 0 and y == 0:
                    continue

                i_edge = i + x
                j_edge = j + y

                if i_edge < 0 or i_edge >= matrix.shape[0] or j_edge < 0 or j_edge >= matrix.shape[1]:
                    continue

                edge_cell = matrix[i_edge, j_edge]

                if edge_cell == 1:
                    neighbors += 1

        return neighbors

    def load_matrix(self):
        with open(self.filepath, 'r') as file:
            rows = [line.strip().split() for line in file]

            self.row_count = len(rows)
            self.column_count = len(rows[0])

            self.matrix = np.zeros(shape=(self.row_count, self.column_count), dtype=np.int32)
            for i in range(self.row_count):
                for j in range(self.column_count):
                    self.matrix[i, j] = int(rows[i][j])

if __name__ == '__main__':
    automata = CellularAutomata('input.txt')

    for i in range(100):
        start = time()
        automata.attribute_next_generation()
        end = time()
        print(f'Generation {i + 1} took {end - start}s')

File: level-ii/question-i/CellularAutomata.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class CellularAutomata {
  private String filepath;
  private List<List<Integer>> matrix;
  private int rowCount;
  private int columnCount;

  public String getFilepath() {
    return filepath;
  }

  public List<List<Integer>> getMatrix() {
    return matrix;
  }

  public int getRowCount() {
    return rowCount;
  }

  public int getColumnCount() {
    return columnCount;
  }

  public CellularAutomata(String filepath) {
    this.filepath = filepath;
    loadMatrix();
  }

  public void attributeNextGeneration() {
    matrix = computeNextGeneration();
  }

  public List<List<Integer>> computeNextGeneration() {
    List<List<Integer>> nextGenerationMatrix = new ArrayList<>(rowCount);
    for (int i = 0; i < rowCount; i++) {
      nextGenerationMatrix.add(new ArrayList<>(columnCount));
      for (int j = 0; j < columnCount; j++) {
        int cell = matrix.get(i).get(j);
        int neighbors = getCellNeighbors(i, j);

        if (cell == 0 && (neighbors > 1 && neighbors < 5)) {
          nextGenerationMatrix.get(i).add(1);
        } else if (cell == 1) {
          if (neighbors > 3 && neighbors < 6) {
            nextGenerationMatrix.get(i).add(1);
          } else {
            nextGenerationMatrix.get(i).add(0);
          }
        } else {
          nextGenerationMatrix.get(i).add(cell);
        }
      }
    }
    return nextGenerationMatrix;
  }

  public int getCellNeighbors(int i, int j) {
    int neighbors = 0;
    for (int x = -1; x < 2; x++) {
      for (int y = -1; y < 2; y++) {
        if (x == 0 && y == 0) {
          continue;
        }

        int iEdge = i + x;
        int jEdge = j + y;

        if (iEdge < 0 || iEdge >= rowCount || jEdge < 0 || jEdge >= columnCount) {
          continue;
        }

        int edgeCell = matrix.get(iEdge).get(jEdge);
        if (edgeCell == 1) {
          neighbors++;
        }
      }
    }
    return neighbors;
  }

  public void loadMatrix() {
    try (BufferedReader br = new BufferedReader(new FileReader(filepath))) {
      List<List<Integer>> rows = new ArrayList<>();
      String line;
      while ((line = br.readLine()) != null) {
        List<Integer> row = new ArrayList<>();
        for (char character : line.toCharArray()) {
          row.add(Character.getNumericValue(character));
        }
        rows.add(row);
      }
      rowCount = rows.size();
      columnCount = rows.get(0).size();
      matrix = new ArrayList<>(rowCount);
      for (int i = 0; i < rowCount; i++) {
        matrix.add(new ArrayList<>(columnCount));
        for (int j = 0; j < columnCount; j++) {
          matrix.get(i).add(rows.get(i).get(j));
        }
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
  }
}
